File: src/test/java/uk/gov/justice/hmpps/casenotes/integration/wiremock/OauthMockServer.java
package uk.gov.justice.hmpps.casenotes.integration.wiremock;

import com.github.tomakehurst.wiremock.junit.WireMockRule;
import uk.gov.justice.hmpps.casenotes.utils.AuthTokenHelper.AuthToken;

import static com.github.tomakehurst.wiremock.client.WireMock.*;

public class OauthMockServer extends WireMockRule {

    private static final int WIREMOCK_PORT = 8998;

    private static final String API_PREFIX = "/auth/api";

    public OauthMockServer() {
        super(WIREMOCK_PORT);
    }

    public void subGetUserDetails(final AuthToken username) {
        stubFor(
                get(urlPathMatching(String.format("%s/user/%s", API_PREFIX, username.name())))
                        .willReturn(aResponse()
                                .withHeader("Content-Type", "application/json")
                                .withBody("{\n" +
                                        "  \"staffId\": 1111,\n" +
                                        "  \"username\": \"" + username.name() + "\",\n" +
                                        "  \"active\": true,\n" +
                                        "  \"name\": \"Mikey Mouse\",\n" +
                                        "  \"authSource\": \"nomis\",\n" +
                                        "  \"activeCaseLoadId\": \"LEI\"\n" +
                                        "}")
                                .withStatus(200)
                        ));

    }

    public void subPing(final int status) {
        stubFor(get("/auth/ping").willReturn(aResponse()
                .withHeader("Content-Type", "application/json")
                .withBody(status == 200 ? "pong" : "some error")
                .withStatus(status)));
    }

}

File: src/test/java/uk/gov/justice/hmpps/casenotes/health/wiremock/OauthMockServer.kt
package uk.gov.justice.hmpps.casenotes.health.wiremock

import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.http.HttpHeader
import com.github.tomakehurst.wiremock.http.HttpHeaders
import org.junit.jupiter.api.extension.AfterAllCallback
import org.junit.jupiter.api.extension.BeforeAllCallback
import org.junit.jupiter.api.extension.BeforeEachCallback
import org.junit.jupiter.api.extension.ExtensionContext

class OAuthExtension :
  BeforeAllCallback,
  AfterAllCallback,
  BeforeEachCallback {
  companion object {
    @JvmField
    val oAuthApi = OAuthMockServer()
  }

  override fun beforeAll(context: ExtensionContext) {
    oAuthApi.start()
    oAuthApi.stubGrantToken()
  }

  override fun beforeEach(context: ExtensionContext) {
    oAuthApi.resetRequests()
    oAuthApi.stubGrantToken()
  }

  override fun afterAll(context: ExtensionContext) {
    oAuthApi.stop()
  }
}

class OAuthMockServer : WireMockServer(WIREMOCK_PORT) {
  fun stubGrantToken() {
    stubFor(
      WireMock.post(WireMock.urlPathMatching(".*"))
        .willReturn(
          WireMock.aResponse()
            .withHeaders(HttpHeaders(HttpHeader("Content-Type", "application/json")))
            .withBody(
              """{
                    "token_type": "bearer",
                    "access_token": "ABCDE"
                }
              """.trimIndent(),
            ),
        ),
    )
  }

  companion object {
    private const val WIREMOCK_PORT = 8998
  }
}
